modin/experimental/engines/omnisci_on_ray/frame/data.py

```diff
@@ -673,6 +689,83 @@ def _union_all(
 
         return new_frame
 
+    def _join_by_index(self, other_modin_frames, how, sort, ignore_index):
+        if how == "outer":
+            raise NotImplementedError("outer join is not supported in OmniSci engine")
+
+        lhs = self._maybe_materialize_rowid()
+        for rhs in other_modin_frames:
+            rhs = rhs._maybe_materialize_rowid()
+            if len(lhs._index_cols) != len(rhs._index_cols):
+                raise NotImplementedError(
+                    "join by indexes with different sizes is not supported"
+                )
+
+            if len(lhs._index_cols) == 1:
+                condition = lhs.ref(lhs._index_cols[0]).eq(rhs.ref(rhs._index_cols[0]))
+            else:
+                condition = OpExpr(
+                    "AND",
+                    [
+                        lhs.ref(lhs_col).eq(rhs.ref(rhs_col))
+                        for lhs_col, rhs_col in zip(lhs._index_cols, rhs._index_cols)
+                    ],
+                    get_dtype(bool),
+                )
+
+            exprs = lhs._index_exprs()
+            new_columns = []
+            for col in lhs.columns:
+                exprs[col] = lhs.ref(col)
+                new_columns.append(col)
+            for col in rhs.columns:
+                # Handle duplicating column names here. When user specifies
+                # suffixes to make a join, actual renaming is done in front-end.
+                new_col_name = col
+                rename_idx = 0
+                while new_col_name in exprs:
+                    new_col_name = f"{col}{rename_idx}"
+                    rename_idx += 1
+                exprs[new_col_name] = rhs.ref(col)
+                new_columns.append(new_col_name)
+
+            op = JoinNode(
+                lhs,
+                rhs,
+                how=how,
+                exprs=exprs,
+                condition=condition,
+            )
+
+            new_columns = Index.__new__(
+                Index, data=new_columns, dtype=self.columns.dtype
+            )
+            lhs = lhs.__constructor__(
```

Comment:
OmniSci itself seems to support joining of multiple tables, at least the expression like the following works at OmniSci:
```sql
SELECT table1.a, table2.b, table3.c FROM table1
JOIN table2 ON table1.id=table2.id 
JOIN table3 ON table1.id=table3.id
```
Is there a way to pack this in a single `JoinNode` expression instead of doing this sequential join in the for-loop?

Reply:
OmniSci requires join node to have exactly two input tables. So, such expression would result in two join operations.